### PR TITLE
[Feature] Page Properties

### LINKED ISSUE #2786: [Feature] Ability to set page order
## Summary
Should be able to set the ordering of pages

## Motivation
Pages can be built in different order from what you want to display. Developer should be able to set an order of page tabs being shown

### PR DESCRIPTION
**Pull request by jsartisan**
Page to manage pages

Fixes #2786

## Test coverage results :test_tube:
<details><summary>:red_circle: Total coverage has decreased</summary>

    // Code coverage diff between base branch:release and head branch: feature/page-properties 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :red_circle: | total | 53.34 **(0)** | 34.84 **(0.08)** | 31.35 **(-0.11)** | 53.89 **(0)**
 :red_circle: | app/client/src/actions/pageActions.tsx | 67.33 **(-0.06)** | 25 **(0)** | 28.57 **(-3.43)** | 56.52 **(-0.62)**
 :red_circle: | app/client/src/api/PageApi.tsx | 64 **(-1.22)** | 25 **(0)** | 16.67 **(-2.08)** | 64.58 **(-1.33)**
 :red_circle: | app/client/src/components/ads/DraggableList.tsx | 19.67 **(0.32)** | 11.11 **(-3.18)** | 0 **(0)** | 24.49 **(-1.04)**
 :green_circle: | app/client/src/constants/DefaultTheme.tsx | 96.33 **(0.03)** | 89.47 **(0)** | 85.71 **(0)** | 95.7 **(0.05)**
 :red_circle: | app/client/src/constants/IconConstants.tsx | 100 **(0)** | 87.5 **(-12.5)** | 100 **(0)** | 100 **(0)**
 :green_circle: | app/client/src/icons/ControlIcons.tsx | 56.91 **(0.81)** | 100 **(0)** | 11.67 **(1.67)** | 56.91 **(0.81)**
 :red_circle: | app/client/src/icons/MenuIcons.tsx | 61.54 **(-0.62)** | 100 **(0)** | 16.67 **(-0.98)** | 61.54 **(-0.62)**
 :green_circle: | app/client/src/pages/Editor/routes.tsx | 41.43 **(0.85)** | 22.22 **(0)** | 6.67 **(0)** | 45.9 **(0.9)**
 :red_circle: | app/client/src/pages/Editor/EditorAppName/NavigationMenuData.ts | 62.16 **(-0.7)** | 12.5 **(0)** | 9.09 **(-0.91)** | 61.11 **(-0.65)**
 :green_circle: | app/client/src/pages/Editor/Explorer/ExplorerIcons.tsx | 87.5 **(0.71)** | 58.33 **(0)** | 20 **(0)** | 87.04 **(0.77)**
 :green_circle: | app/client/src/pages/Editor/Explorer/Entity/AddButton.tsx | 87.5 **(2.88)** | 62.5 **(-4.17)** | 66.67 **(16.67)** | 85.71 **(3.89)**
 :green_circle: | app/client/src/pages/Editor/Explorer/Entity/CollapseToggle.tsx | 92.31 **(0)** | 83.33 **(16.66)** | 100 **(0)** | 100 **(0)**
 :green_circle: | app/client/src/pages/Editor/Explorer/Entity/Name.tsx | 58.9 **(0.46)** | 27.5 **(1.31)** | 45.45 **(6.99)** | 57.97 **(-0.14)**
 :red_circle: | app/client/src/pages/Editor/Explorer/Entity/index.tsx | 92.96 **(-2.2)** | 79.63 **(-0.8)** | 85 **(-8.75)** | 92.65 **(-2.27)**
 :green_circle: | app/client/src/pages/Editor/Explorer/Pages/PageContextMenu.tsx | 75 **(3.57)** | 60 **(0)** | 16.67 **(4.17)** | 77.42 **(3.89)**
 :red_circle: | app/client/src/pages/Editor/Explorer/Pages/PageGroup.tsx | 78.95 **(-3.4)** | 58.82 **(0)** | 50 **(-16.67)** | 81.08 **(-3.77)**
 :sparkles: :new: | **app/client/src/pages/Editor/PagesEditor/ContextMenu.tsx** | **60** | **60** | **0** | **63.64**
 :sparkles: :new: | **app/client/src/pages/Editor/PagesEditor/EditName.tsx** | **37.84** | **20** | **0** | **42.42**
 :sparkles: :new: | **app/client/src/pages/Editor/PagesEditor/PageListItem.tsx** | **58.33** | **55.56** | **0** | **58.33**
 :sparkles: :new: | **app/client/src/pages/Editor/PagesEditor/index.tsx** | **50** | **60** | **0** | **51.16**
 :red_circle: | app/client/src/reducers/entityReducers/pageListReducer.tsx | 17.65 **(-0.13)** | 25 **(0)** | 15 **(-2.65)** | 20 **(0)**
 :red_circle: | app/client/src/sagas/PageSagas.tsx | 23.44 **(-0.51)** | 6.25 **(-0.51)** | 14.29 **(-0.42)** | 25.18 **(-0.48)**
 :red_circle: | app/client/src/utils/autocomplete/TernServer.ts | 50.47 **(-0.24)** | 37.72 **(-0.88)** | 36.21 **(0)** | 55.08 **(-0.27)**</details>

### CODE REVIEW

**Review by riodeuno — APPROVED** [review]
Approve for tests

File: app/client/src/pages/Editor/Explorer/Pages/PageGroup.tsx [hunk 1]
```diff
@@ -80,12 +82,21 @@ export const ExplorerPageGroup = memo((props: ExplorerPageGroupProps) => {
 
   return (
     <Entity
+      action={() =>
+        history.push(PAGE_LIST_EDITOR_URL(params.applicationId, params.pageId))
```
Comment (riodeuno) [inline app/client/src/pages/Editor/Explorer/Pages/PageGroup.tsx:86]: Let's useCallback

File: app/client/src/pages/Editor/Explorer/Pages/PageGroup.tsx [hunk 2]
```diff
@@ -80,12 +82,21 @@ export const ExplorerPageGroup = memo((props: ExplorerPageGroupProps) => {
...
+      }
+      alwaysShowRightIcon
       className="group pages"
+      disabled
       entityId="Pages"
       icon={pageGroupIcon}
       isDefaultExpanded
       name="Pages"
+      onClickRightIcon={() => {
+        history.push(PAGE_LIST_EDITOR_URL(params.applicationId, params.pageId));
```
Comment (riodeuno) [inline app/client/src/pages/Editor/Explorer/Pages/PageGroup.tsx:96]: Let's useCallback

File: app/client/src/pages/Editor/PagesEditor/ContextMenu.tsx [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+      content={
+        <Container>
+          <Header>
+            <PageName>
+              <EditName applicationId={applicationId} page={page} />
+            </PageName>
+            <Actions>
+              <Action>
+                <CopyIcon
+                  color={get(theme, "colors.propertyPane.iconColor")}
```
Comment (riodeuno) [inline app/client/src/pages/Editor/PagesEditor/ContextMenu.tsx:123]: Let's make it a constant, or in a `useMemo`

File: app/client/src/pages/Editor/PagesEditor/ContextMenu.tsx [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+                  height={16}
+                  onClick={() => onCopy(page.pageId)}
```
Comment (riodeuno) [inline app/client/src/pages/Editor/PagesEditor/ContextMenu.tsx:125]: Let's useCallback

File: app/client/src/pages/Editor/PagesEditor/ContextMenu.tsx [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+  } = props;
+  const theme = useTheme();
+  const [isOpen, setIsOpen] = useState(false);
+
+  const handleInteraction = useCallback((isOpen) => {
+    setIsOpen(isOpen);
+  }, []);
+
+  return (
+    <Popover2
```
Comment (riodeuno) [inline app/client/src/pages/Editor/PagesEditor/ContextMenu.tsx:113]: Lets split Popover2 as an appsmith component. Also, content should be a new memoized component

File: app/client/src/pages/Editor/PagesEditor/ContextMenu.tsx [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+                <div>Set Homepage</div>
+                <MenuItemToggle
+                  onToggle={() => onSetPageDefault(page.pageId, applicationId)}
+                  value={page.isDefault}
+                />
+              </MenuItem>
+            )}
+
+            <MenuItem>
+              <div>IsVisible</div>
```
Comment (riodeuno) [inline app/client/src/pages/Editor/PagesEditor/ContextMenu.tsx:169]: Let's follow the labels we have in the property pane.

File: app/client/src/pages/Editor/PagesEditor/ContextMenu.tsx [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+      minimal
+      onInteraction={handleInteraction}
+      placement={"bottom-start"}
+      portalClassName="pages-editor-context-menu"
+    >
+      <Action className={isOpen ? "active" : ""} type="button">
+        <SettingsIcon
+          color={get(theme, "colors.propertyPane.iconColor")}
+          height={16}
+          onClick={() => {
```
Comment (riodeuno) [inline app/client/src/pages/Editor/PagesEditor/ContextMenu.tsx:188]: Let's remove this if it is not needed. or we can use `noop`

File: app/client/src/pages/Editor/PagesEditor/ContextMenu.tsx [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+                value={!page.isHidden}
+              />
+            </MenuItem>
+          </main>
+        </Container>
+      }
+      isOpen={isOpen}
+      minimal
+      onInteraction={handleInteraction}
+      placement={"bottom-start"}
```
Comment (riodeuno) [inline app/client/src/pages/Editor/PagesEditor/ContextMenu.tsx:181]: We don't need `{}` for a string prop

File: app/client/src/pages/Editor/PagesEditor/index.tsx [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+        />
+      </Header>
+
+      <DraggableList
+        ItemRenderer={({ item }: any) => (
+          <PageListItem applicationId={params.applicationId} item={item} />
+        )}
+        itemHeight={70}
+        items={pages}
+        onUpdate={(newOrder: any, originalIndex: number, newIndex: number) => {
```
Comment (riodeuno) [inline app/client/src/pages/Editor/PagesEditor/index.tsx:151]: Let's see if we need useCallback here.

File: app/client/src/sagas/PageSagas.tsx [hunk 10]
```diff
@@ -111,6 +112,7 @@ export function* fetchPageListSaga(
     const isValidResponse = yield validateResponse(response);
     if (isValidResponse) {
       const orgId = response.data.organizationId;
+      console.log({ pages: response.data });
```
Comment (riodeuno) [inline app/client/src/sagas/PageSagas.tsx:115]: missed console.log

**Review by riodeuno — CHANGES_REQUESTED** [review]

**Comment by somangshu** [comment]
@jsartisan can you link the issue to this PR?

**Comment by marks0351** [comment]
@jsartisan  deleting from the context menu works fine but doesn't seem to work on the page list item. code looks good.

**Comment by marks0351** [comment]
> @jsartisan deleting from the context menu works fine but doesn't seem to work on the page list item. code looks good.

my bad forgot to move the netlify draw out of the way. works fine!!

**Comment by jsartisan** [comment]
@vivekverma2312 this PR is ready for QA.

**Comment by vivekverma2312** [comment]
@jsartisan 
1. Home page can be changed to invisible and deploy which launch blank app
https://www.loom.com/share/473b0fdf72d04ca3a70aa12d0053f903
2. Not able to move page at bottom due to page scroll on page orderhttps://www.loom.com/share/fee3e22afb054c7b86eb9104b5c5379c
3. Unable to find page error on deploy
Screenshot 2021-08-09 at 1 35 02 PM
[app](https://deploy-preview-4989--appsmith-editor-ce.netlify.app/applications/6110d46cfd54bb5f65658ca0/pages/6110d46cfd54bb5f65658ca2) for repro
4. Page1 is set as Homepage but post deployment page 2 is selected by default
https://www.loom.com/share/7008eec025a04b4cbaaba5e575abea2f

5. Visible property is available as per mock it is invisible with JS option
6. URL property is not available on property pane
for 5 and 6  -- 
Screenshot 2021-08-09 at 1 21 11 PM

**Comment by vivekverma2312** [comment]
@jsartisan This looks fine. For issue2 I will test in different PR. Good to be merged in release.

**Pull request merged**